pybamm/expression_tree/concatenations.py

```diff
@@ -73,6 +73,114 @@ def __init__(self, *children):
 
         super().__init__(*children, name="numpy concatenation")
 
-    def evaluate(self, t, y):
+    def evaluate(self, t=None, y=None):
         """ See :meth:`pybamm.Symbol.evaluate()`. """
         return np.concatenate([child.evaluate(t, y) for child in self.children])
+
+
+class DomainConcatenation(Concatenation):
+    """A node in the expression tree representing a concatenation of symbols.
+
+    It is assumed that each child has a domain, and the final concatenated vector will
+    respect the sizes and ordering of domains established in pybamm.KNOWN_DOMAINS
+
+    **Extends**: :class:`pybamm.Concatenation`
+
+    Parameters
+    ----------
+
+    children : iterable of :class:`pybamm.Symbol`
+        The symbols to concatenate
+
+    mesh : :class:`pybamm.BaseMesh` (or subclass)
+        The underlying mesh for discretisation, used to obtain the number of mesh points
+        in each domain.
+
+    """
+
+    def __init__(self, children, mesh):
+        # Convert any constant symbols in children to a Vector of the right size for
+        # concatenation
+
+        children = list(children)
+
+        for i, child in enumerate(children):
+            if child.is_constant():
+                children[i] = self.process_node_for_concantate(child, mesh)
+
+        # Allow the base class to sort the domains into the correct order
+        super().__init__(*children, name="numpy concatenation")
+
+        # deal with "whole cell" special case
+        if self.domain == ["whole cell"]:
+            self.domain = ["negative electrode", "separator", "positive electrode"]
+
+        # create dict of domain => slice of final vector
+        self._slices = self.create_slices(self, mesh)
+
+        # store size of final vector
+        self._size = self._slices[self.domain[-1]].stop
+
+        # deal with "whole cell" special case
```

Comment:
duplicate?

Reply:
good catch, that was a copy and paste I forgot to change! should be fixed now